Floating Point Error Estimation RST added (#648)
vgvassilev/clad: docs/userDocs/source/user/FloatingPointErrorEstimation.rst

FP Error Estimation
*******************

============
Introduction
============

Data-intensive sciences that work with increasing data volumes and often in
heterogeneous computing environments require floating point stability. Robust
floating-point error detection can help reduce data reprocessing costs and help
develop important new lossy compression algorithms.

How does Automatic Differentiation (AD) fit into this?
======================================================

AD helps evaluate the exact derivative of a function. AD applies the chain rule
of differential calculus throughout the semantics of the original program. In

> the context of FP error estimation, this research uses:

Reviewer: Can you please clarify what do you mean by 'this research uses'?

Author: Thanks, I just realized that the phrasing is a bit off... Rephrasing as suggested by Vassil below:

AD helps evaluate the exact derivative of a function. AD applies the
**differential calculus chain rule** throughout the semantics of the original
program. In the context of FP error estimation, **~this research~ the implementation relies on:**:

- **Reverse-Mode AD** (as opposed to Forward-Mode AD), since it provides the
  derivative of the function with respect to all intermediate and input
  variables, and

- **Source Transformation** (as opposed to Operator Overloading), since it does
  most of the work at compile time. The Clad Framework that is used in this
  research also uses source transformation.

   **Clad** is implemented as a plugin for the Clang compiler. It inspects the
   internal compiler representation of the target function to generate its
   derivative. Clad requires little or no code modification, supports a growing
   subset of C++ constructs, statements and data types, it enables efficient
   gradient computation for large and complex codebases, and is deeply
   integrated with the compiler, allowing automatic generation of error
   estimation code.

Where does the FPEE logic reside?
=================================

> The main logic for Floating Point Error Estimation resides in the following

Reviewer: We should not add implementation details in the user documentation. For user documentation, we should perhaps focus on these things:

- What is error estimation? 
- Why use AD in error estimation as opposed to other differentiation techniques in error estimation?
- What will happen if AD fails?
- Tutorials/Examples that show how to use Clad's error estimation functionality effectively.

@vgvassilev and @grimmmyshini can comment better on the documentation structure.

Author: I have linked the Doxygen and ReadMe docs in the comment below

files:

- `include/clad/Differentiator/ErrorEstimator.h`_

  - uses ``ErrorEstimationHandler`` class to estimate errors in a target
    function.  It keeps track of error expressions, emits error statements, and
    replaces parameter values.

- `include/clad/Differentiator/EstimationModel.h`_

  - contains the information needed to calculate the estimate value of the
    error.

   Above files include a lot of useful documentation in the form of code
   comments. 

How does the FPEE Logic work?
=============================

While parsing the code using Clad, if it encounters a floating point

> variable,it needs to be registered into the system (to accumulate relevant

Reviewer: What is system here?

Author: Rephrased as suggested by Vassil:

While parsing the code using Clad, if it encounters a floating point variable,
it needs to be **tracked**

> variable,it needs to be registered into the system (to accumulate relevant

Reviewer 2: ```suggestion
variable, it needs to be tracked (to accumulate relevant
```

> errors against that variable). Next, the Error Estimation Calcualtion Formula

Reviewer: ```suggestion
errors against that variable). Next, the Error Estimation Calculation Formula
```

(Error Model) needs to be built (using ``EstimationModel.h``).

``EstimationModel.h`` contains the information needed to calculate the estimate 
value of the error. It is highly customizable (e.g., you can plug in your 
own custom formula as well). The default formula multiplies the derivative 
(dfdx) with the value of the variable (delta_x), for which the error estimate 
is required, and the machine epsilon (Em).

``std::abs(dfdx * delta_x * Em)``

  For this formula to work, the value of the variable (delta_x) should be saved
  at the relevant time.

> This model will return a formula that is represented using a Clang

Reviewer: Users perhaps do not care at all if the error is a clang expression. We should write this documentation in a way that is useful for the users.

Author: This is as described by Garima. As I understand, the idea was that familiarity with Clang expressions is required, since we're targeting a dev audience, including new comers... Please let me know if you have an alternate text suggestion

> FP Error Estimation

Reviewer 3: Don't abbreviate the title.. Write out Floating Point

Reviewer 3: Branding - this tool is called CHEF-FP. We should refer to it that way (repeatedly)

function. When you use Clad in Error Estimation mode, the
``ErrorEstimationHandler`` class is responsible for handling the derivative and
error information exchange between Clad and the Error Estimation module
(CHEF-FP). This class is responsible for a lot of the housekeeping tasks as
well.

  **CHEF-FP** is a flexible, scalable, and easy-to-use source-code
  transformation tool based on Automatic Differentiation (AD) for analyzing
  approximation errors in HPC applications. For more details, please view `Fast
  and Automatic Floating Point Error Analysis`_.

clad::ErrorEstimator::EmitFinalErrorStmts
-----------------------------------------

This function adds the final error and the other parameter errors to the
forward block.

.. _include/clad/Differentiator/ErrorEstimator.h: https://github.com/vgvassilev/clad/blob/master/include/clad/Differentiator/ErrorEstimator.h

.. _include/clad/Differentiator/EstimationModel.h: https://github.com/vgvassilev/clad/blob/master/include/clad/Differentiator/EstimationModel.h

.. _demos/ErrorEstimation: https://github.com/vgvassilev/clad/tree/master/demos/ErrorEstimation

> .. _Fast and Automatic Floating Point Error Analysis: https://arxiv.org/pdf/2304.06441.pdf

Reviewer 3: use complete title

FP Error Estimation
*******************

============
Introduction
============

Data-intensive sciences that work with increasing data volumes and often in
heterogeneous computing environments require floating point stability. Robust
floating-point error detection can help reduce data reprocessing costs and help
develop important new lossy compression algorithms.

> Who is this article for?

Reviewer 2: Perhaps we can remove this section?

Author: section removed and commit squashed to existing commit
